// apollo-language-server/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Debug = exports.graphqlTypes = exports.schemaProviderFromConfig = exports.GraphQLServiceProject = exports.isServiceProject = exports.GraphQLClientProject = exports.isClientProject = exports.GraphQLProject = exports.ToolError = exports.getValidationErrors = void 0;
var validation_1 = require("./errors/validation");
Object.defineProperty(exports, "getValidationErrors", { enumerable: true, get: function () { return validation_1.getValidationErrors; } });
var logger_1 = require("./errors/logger");
Object.defineProperty(exports, "ToolError", { enumerable: true, get: function () { return logger_1.ToolError; } });
var base_1 = require("./project/base");
Object.defineProperty(exports, "GraphQLProject", { enumerable: true, get: function () { return base_1.GraphQLProject; } });
var client_1 = require("./project/client");
Object.defineProperty(exports, "isClientProject", { enumerable: true, get: function () { return client_1.isClientProject; } });
Object.defineProperty(exports, "GraphQLClientProject", { enumerable: true, get: function () { return client_1.GraphQLClientProject; } });
var service_1 = require("./project/service");
Object.defineProperty(exports, "isServiceProject", { enumerable: true, get: function () { return service_1.isServiceProject; } });
Object.defineProperty(exports, "GraphQLServiceProject", { enumerable: true, get: function () { return service_1.GraphQLServiceProject; } });
var schema_1 = require("./providers/schema");
Object.defineProperty(exports, "schemaProviderFromConfig", { enumerable: true, get: function () { return schema_1.schemaProviderFromConfig; } });
__exportStar(require("./engine"), exports);
__exportStar(require("./config"), exports);
const graphqlTypes = __importStar(require("./graphqlTypes"));
exports.graphqlTypes = graphqlTypes;
var utilities_1 = require("./utilities");
Object.defineProperty(exports, "Debug", { enumerable: true, get: function () { return utilities_1.Debug; } });
